File: day10/main.go
package main

import (
	"bufio"
	"errors"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatalln(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatalln(err)
		}
	}()

	vm := NewVM()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.Fields(scanner.Text())
		instr := line[0]
		arg := 0
		if len(line) > 1 {
			var err error
			arg, err = strconv.Atoi(line[1])
			if err != nil {
				log.Fatalln(err)
			}
		}
		vm.Exec(instr, arg)
	}

	if err := scanner.Err(); err != nil {
		log.Fatalln(err)
	}

	fmt.Println("Part 1:", vm.strength)
	fmt.Println("Part 2:")
	for _, i := range vm.grid {
		fmt.Println(string(i))
	}
}

type (
	VM struct {
		cycle       int
		rx          int
		strength    int
		cycleTarget int
		cycleIncr   int
		grid        [][]byte
		scanline    int
	}
)

const (
	screenWidth  = 40
	screenHeight = 6
	screenSize   = screenWidth * screenHeight
)

func NewVM() *VM {
	grid := make([][]byte, screenHeight)
	for i := range grid {
		grid[i] = make([]byte, screenWidth)
	}
	return &VM{
		cycle:       0,
		rx:          1,
		strength:    0,
		cycleTarget: 20,
		cycleIncr:   40,
		grid:        grid,
		scanline:    0,
	}
}

func (m *VM) Exec(instr string, arg int) error {
	switch instr {
	case "noop":
		m.execNoop()
	case "addx":
		m.execAddX(arg)
	default:
		return errors.New("Invalid instruction")
	}
	return nil
}

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}

func (m *VM) intersect(x int) bool {
	return abs(x-m.rx) <= 1
}

func (m *VM) checkCycle(cycle int, set bool, next int) {
	m.cycle += cycle
	if m.cycle >= m.cycleTarget {
		m.strength += m.cycleTarget * m.rx
		m.cycleTarget += m.cycleIncr
	}
	y := m.scanline / screenWidth
	x := m.scanline % screenWidth
	if m.intersect(x) {
		m.grid[y][x] = '#'
	} else {
		m.grid[y][x] = '.'
	}
	m.scanline = (m.scanline + 1) % screenSize
	if set {
		m.rx = next
	}
}

func (m *VM) execNoop() {
	m.checkCycle(1, false, 0)
}

func (m *VM) execAddX(arg int) {
	m.checkCycle(1, false, 0)
	m.checkCycle(1, true, m.rx+arg)
}

File: day10/src/main.rs
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

type BoxError = Box<dyn std::error::Error>;
type BoxResult<T> = Result<T, BoxError>;

fn main() -> BoxResult<()> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let mut vm = VM::new();

    for line in reader.lines() {
        let line = line?;
        let (instr, arg) = match line.split_ascii_whitespace().collect::<Vec<_>>()[..] {
            [instr] => (instr, 0),
            [instr, arg] => (instr, arg.parse::<i32>()?),
            _ => return Err("Invalid instruction line".into()),
        };
        vm.exec(instr, arg)?;
    }

    println!("Part 1: {}\nPart 2:", vm.strength);
    for line in vm.grid {
        println!("{}", String::from_utf8(line)?);
    }

    Ok(())
}

const SCREEN_WIDTH: usize = 40;
const SCREEN_HEIGHT: usize = 6;
const SCREEN_SIZE: usize = SCREEN_WIDTH * SCREEN_HEIGHT;

struct VM {
    cycle: i32,
    rx: i32,
    strength: i32,
    cycle_target: i32,
    cycle_incr: i32,
    grid: Vec<Vec<u8>>,
    scanline: usize,
}

impl VM {
    fn new() -> Self {
        Self {
            cycle: 0,
            rx: 1,
            strength: 0,
            cycle_target: 20,
            cycle_incr: 40,
            grid: vec![vec![0; SCREEN_WIDTH]; SCREEN_HEIGHT],
            scanline: 0,
        }
    }

    fn exec(&mut self, instr: &str, arg: i32) -> BoxResult<()> {
        match instr {
            "noop" => self.exec_noop(),
            "addx" => self.exec_addx(arg),
            _ => return Err("Invalid instruction".into()),
        }
        Ok(())
    }

    fn intersect(&self, x: i32) -> bool {
        (x - self.rx).abs() <= 1
    }

    fn check_cycle(&mut self, cycle: i32, set: bool, next: i32) {
        self.cycle += cycle;
        if self.cycle >= self.cycle_target {
            self.strength += self.cycle_target * self.rx;
            self.cycle_target += self.cycle_incr;
        }
        let y = self.scanline / SCREEN_WIDTH;
        let x = self.scanline % SCREEN_WIDTH;
        self.grid[y][x] = if self.intersect(x as i32) { b'#' } else { b'.' };
        self.scanline = (self.scanline + 1) % SCREEN_SIZE;
        if set {
            self.rx = next;
        }
    }

    fn exec_noop(&mut self) {
        self.check_cycle(1, false, 0)
    }

    fn exec_addx(&mut self, arg: i32) {
        self.check_cycle(1, false, 0);
        self.check_cycle(1, true, self.rx + arg)
    }
}
